src/checkpoint.hh

```diff
@@ -0,0 +1,140 @@
+
+namespace Printer {
+class Checkpoint : public Base {
+ public:
+    size_t interval = 0;
+
+    explicit Checkpoint(size_t interval): interval(interval) {}
+
+    void include(Printer::Base &stream) {
+      stream << "#include \"boost/serialization/vector.hpp\"" << endl;
+      stream << "#include \"boost/archive/binary_iarchive.hpp\"" << endl;
+      stream << "#include \"boost/archive/binary_oarchive.hpp\"" << endl;
+      stream << "#include <atomic>" << endl;
+      stream << "#include <thread>" << endl << endl;
+    }
+
+    void archive(Printer::Base &stream) {
+      inc_indent(); inc_indent();
+      stream << indent();
+      stream << "void archive(const std::string &output_file) {" << endl;
+      inc_indent();
+      stream << indent() << "// save the DP table/array to disk" << endl;
+      stream << indent() << "try {" << endl;
+      inc_indent();
+      stream << indent() << "std::ofstream array_fout(output_file, "
+                            "std::ios::binary);" << endl;
+      stream << indent() << "boost::archive::binary_oarchive "
+                            "array_out(array_fout);" << endl;
+      stream << indent() << "array_out << array;" << endl;
+      stream << indent() << "array_fout.close();" << endl;
+      dec_indent();
+      stream << indent() << "} catch (std::exception &e) {" << endl;
+      inc_indent();
+      stream << indent() << "std::cerr << \"Error during archiving of \\\"\" "
+                            "<< output_file << \"\\\" table!\""
+             << endl;
+      stream << indent() << "          << \" Will retry at "
+                            "the next checkpoint...\\n\";" << endl;
+      dec_indent();
+      stream << indent() << "}" << endl;
+      dec_indent();
+      stream << indent() << "}" << endl << endl;
+      dec_indent(); dec_indent();
+    }
+
+    void init(Printer::Base &stream) {
+      inc_indent(); inc_indent(); inc_indent();
+      stream << indent() << "// read the DP array/table from disk "
+                            "and put its contents into array" << endl;
+      stream << indent() << "try {" << endl;
+      inc_indent();
+      stream << indent() << "std::ifstream array_fin(tname, "
+                            "std::ios::binary);" << endl;
+      stream << indent() << "boost::archive::binary_iarchive "
+                            "array_in(array_fin);" << endl;
+      stream << indent() << "array_in >> array;" << endl;
+      stream << indent() << "array_fin.close();" << endl;
+      stream << indent() << "// mark the already existing table values "
+                            "in the tabulated vector" << endl;
+      stream << indent() << "for (long unsigned int i = 0; i < array.size(); "
+                            "i++) tabulated[i] = array[i];" << endl;
+      dec_indent();
+      stream << indent() << "} catch (const std::exception &e) {" << endl;
+      inc_indent();
+      stream << indent() << "std::cerr << \"Error \\\"\" << e.what() << \"\\\" "
```

Comment:
This error message might scare the user too much especially when executing the binary for the first time:
```(base) t490s x86_64 /Daten/Git/jlab/gapc>./out_cp uagucguuagcugggacuugaucguagucguuagcugggacuugaucguagucguuagcugggacuugaucguagucguuagcugggacuugaucguagucguuagcugggacuugaucguagucguuagcugggacuugaucguagucguuagcugggacuugaucguagucguuagcugggacuugaucg
Error "input stream error" while reading "dangle_table" table! This table will be initialized empty!
Error "input stream error" while reading "iloop_table" table! This table will be initialized empty!
Error "input stream error" while reading "ml_comps_table" table! This table will be initialized empty!
Error "input stream error" while reading "ml_comps1_table" table! This table will be initialized empty!
Error "input stream error" while reading "strong_table" table! This table will be initialized empty!
Error "input stream error" while reading "struct_table" table! This table will be initialized empty!
Error "input stream error" while reading "weak_table" table! This table will be initialized empty!
Answer: 
-8790
```

as this is the correct behavior if no check point has been dumped yet

Reply:
Agreed. I guess it would maybe be better to print a different message when no checkpoints have been created yet. I will think of something better..